extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRuntimeClient.java

```diff
@@ -0,0 +1,238 @@
+
+        auth.authenticate(params, new Handler<AsyncResult<User>>() {
+            @Override
+            public void handle(AsyncResult<User> result) {
+                if (result.succeeded()) {
+                    final AccessToken token = AccessToken.class.cast(result.result());
+                    final String opaqueIdToken = token.opaqueIdToken();
+                    final String opaqueAccessToken = token.opaqueAccessToken();
+                    final String opaqueRefreshToken = token.opaqueRefreshToken();
+
+                    AuthorizationCodeTokens tokens = new AuthorizationCodeTokens(opaqueIdToken, opaqueAccessToken,
+                            opaqueRefreshToken);
+                    emitter.complete(tokens);
+                } else {
+                    emitter.fail(result.cause());
+                }
+            }
+        });
+    }
+
+    public String getLogoutPath() {
+        return OAuth2AuthProviderImpl.class.cast(auth).getConfig().getLogoutPath();
+    }
+
+    public void decodeToken(String token, Handler<AsyncResult<AccessToken>> resultHandler) {
+        auth.decodeToken(token, resultHandler);
+    }
+
+    public void createUserInfoToken(UniEmitter<? super JsonObject> uniEmitter, RoutingContext vertxContext,
+            TokenAuthenticationRequest request) {
+        OAuth2TokenImpl tokenImpl = new OAuth2TokenImpl(auth, new JsonObject());
+        String accessToken = vertxContext.get("access_token");
+        if (accessToken == null) {
+            accessToken = request.getToken().getToken();
+        }
+        tokenImpl.principal().put("access_token", accessToken);
+        tokenImpl.userInfo(new Handler<AsyncResult<JsonObject>>() {
+            @Override
+            public void handle(AsyncResult<JsonObject> event) {
+                if (event.failed()) {
+                    uniEmitter.fail(new AuthenticationFailedException(event.cause()));
+                } else {
+                    uniEmitter.complete(event.result());
+                }
+            }
+        });
+    }
+
+    public static OidcRuntimeClient discoverOidcEndpoints(Vertx vertx, OAuth2ClientOptions options,
+            OidcTenantConfig oidcConfig) {
+        return Uni.createFrom().emitter(new Consumer<UniEmitter<? super OidcRuntimeClient>>() {
+            public void accept(UniEmitter<? super OidcRuntimeClient> uniEmitter) {
+                KeycloakAuth.discover(vertx, options, new Handler<AsyncResult<OAuth2Auth>>() {
+                    @Override
+                    public void handle(AsyncResult<OAuth2Auth> event) {
+                        if (event.failed()) {
+                            uniEmitter.fail(toOidcException(event.cause(), options.getSite()));
+                        } else {
+                            uniEmitter.complete(createClient(event.result(), oidcConfig));
+                        }
+                    }
+                });
+            }
+        }).await().indefinitely();
```

Comment:
That will be problematic

1. it may block forever
2. you may block the wrong thread 

Is it possible to return a `Uni<OidcRuntimeClient>` ?